velox/functions/sparksql/tests/StringTest.cpp

```diff
@@ -370,6 +378,34 @@ TEST_F(StringTest, endsWith) {
   EXPECT_EQ(endsWith(std::nullopt, "abc"), std::nullopt);
 }
 
+TEST_F(StringTest, substringIndex) {
+  EXPECT_EQ(substring_index("www.apache.org", ".", 3), "www.apache.org");
+  EXPECT_EQ(substring_index("www.apache.org", ".", 2), "www.apache");
+  EXPECT_EQ(substring_index("www.apache.org", ".", 1), "www");
+  EXPECT_EQ(substring_index("www.apache.org", ".", 0), "");
+  EXPECT_EQ(substring_index("www.apache.org", ".", -1), "org");
+  EXPECT_EQ(substring_index("www.apache.org", ".", -2), "apache.org");
+  EXPECT_EQ(substring_index("www.apache.org", ".", -3), "www.apache.org");
+  // Str is empty string.
+  EXPECT_EQ(substring_index("", ".", 1), "");
+  // Empty string delim.
+  EXPECT_EQ(substring_index("www.apache.org", "", 1), "");
+  // Delim does not exist in str.
+  EXPECT_EQ(substring_index("www.apache.org", "#", 2), "www.apache.org");
+  EXPECT_EQ(substring_index("www.apache.org", "WW", 1), "www.apache.org");
+  // Delim is 2 chars.
+  EXPECT_EQ(substring_index("www||apache||org", "||", 2), "www||apache");
+  EXPECT_EQ(substring_index("www||apache||org", "||", -2), "apache||org");
+  // Non ascii chars.
+  EXPECT_EQ(substring_index("大千世界大千世界", "千", 2), "大千世界大");
```

Comment:
Maybe add a test for invalid UTF-8 string.

Reply:
Added.